Add retention period to all SIP origin configs

jraddaoui opened the pull request:
Allow setting a retention period to the internal uploads and the SIP
sources, and update the current behavior in watched buckets for
consistency. The retention period will default to "0" (automatic 
deletion) and has to be set to a negative value to disable the deletion
of the original SIP after a successful ingest. Record a workflow task to
show details in the API/dashboard.

Refs #1355.

jraddaoui committed:
Generate mocks after mockgen upgrade

jraddaoui commented:
Still working on the tests and the documentation, but this is ready for an initial review when you have some time, @djjuhasz!

djjuhasz commented on cmd/enduro-a3m-worker/main.go:
```diff
@@ -247,7 +247,8 @@ func main() {
 		// work similar to the watched bucket download activity and use the source ID
-		// to determine which bucket to use.
+		// to determine which bucket to use. Alternatively, we could register multiple
+		// copies of this activity, one per source.
```
How would registering the activity multiple times work? Is it enough to register the activity with a different `Name` each time? :thinking:

jraddaoui replied:
Yeah, we already do that:

https://github.com/artefactual-sdps/enduro/blob/main/cmd/enduro-am-worker/main.go#L258-L265

It's the same activity from temporal-activities, we just register them with different names and buckets.

djjuhasz commented on enduro.toml:
```diff
@@ -132,6 +132,11 @@ bucket = "sips"
+# successful ingest. Do not set to disable automatic deletion of SIPs. Set to "0"
+# to delete uploaded SIPs immediately after they have been ingested.
+# It must be a string format compatible with https://pkg.go.dev/time#ParseDuration.
+retentionPeriod = "0"
```
I think it would be better to use a negative duration for indefinite storage so a zero or positive value in this config file can be overridden by an environment variable that disables deletion (e.g. `ENDURO_WATCHER_EMBEDDED_RETENTIONPERIOD=-1s`).

jraddaoui replied:
It was already like this and that's quite an edge case, but we can change it.

jraddaoui replied:
I like the idea of being unset means no deletion, in case somebody forgets it. We could keep both options if you want ("Do not set or use a negative value to disable automatic deletion of SIPs.") but I think it's more confusing just to support environment overrides. I'd keep it as it is unless you feel strongly about it.

djjuhasz replied:
Defaulting to keeping SIPs forever makes sense to me.  We might want to change the comment wording a bit to make it very clear that the `retentionPeriod` setting should be commented out or removed entirely to disable automatic deletion.

djjuhasz replied:
Oh okay, I'd forgotten we decided to make the default behaviour deleting the SIP immediately.  Ignore the above.

djjuhasz commented on internal/workflow/delete_original.go:
```diff
@@ -0,0 +1,92 @@
+		ctx,
+		&datatypes.Task{
+			Name:         "Delete original SIP",
+			Note:         fmt.Sprintf("The original SIP will be deleted in %s", state.req.RetentionPeriod.String()),
```
It might be more useful to report the time at which the SIP will be deleted, rather than the duration.

jraddaoui replied:
But being part of the note field it's not going to work across different timezones. I thought we displayed the start time in the tasks while they are in progress, but I don't see that now in the dashboard. I think that may be a better solution.

djjuhasz replied:
Ah, good point about the timezone problems with displaying the deletion time. I agree that it would be helpful to show the start time for a task in progress.

djjuhasz commented on internal/workflow/delete_original.go:
```diff
@@ -0,0 +1,92 @@
+	}
+
+	// Set a timer for the retention period.
+	if err := temporalsdk_workflow.NewTimer(ctx, *state.req.RetentionPeriod).Get(ctx, nil); err != nil {
```
I think you can use `temporalsdk_workflow.Sleep(ctx, *state.req.RetentionPeriod)` to block until the timer is complete, `NewTimer()` looks to me like it's meant more for asynchronous use.

jraddaoui replied:
Looks like Sleep uses the timer internally, so just a little cleaner for us.

https://github.com/temporalio/sdk-go/blob/master/internal/workflow.go#L1556-L1574

djjuhasz commented on internal/workflow/processing.go:
```diff
@@ -246,31 +246,23 @@ func (w *ProcessingWorkflow) Execute(ctx temporalsdk_workflow.Context, req *inge
+	if state.status == enums.WorkflowStatusDone {
+		if req.RetentionPeriod != nil {
+			if err := w.deleteOriginalSIP(ctx, state); err != nil {
+				w.logger.Error("Failed to delete original SIP", "err", err.Error())
```
The error returned from `w.deleteOriginalSIP()` includes "delete original SIP" in the message too, I'd recommend returning the deletion error without the prefix text from `w.deleteOriginalSIP()`.

djjuhasz commented on internal/workflow/processing.go:
```diff
@@ -246,31 +246,23 @@ func (w *ProcessingWorkflow) Execute(ctx temporalsdk_workflow.Context, req *inge
-					req.Key,
-				).Get(activityOpts, nil)
+	// Schedule deletion or disposal of the original SIP.
+	if state.status == enums.WorkflowStatusDone {
```
This multi-nested conditional is hard to understand.  Can we just return here if the workflow status is not done? e.g. 

```
if state.status != enums.WorkflowStatusDone {
    return nil
}
```

The only thing we'd be skipping that runs now is the final "Workflow completed successfully!" log message, but I don't think that makes sense for a status other than "done".  I'm not sure we can even reach this point in the workflow without `state.status == enums.WorkflowStatusDone`. :thinking:

jraddaoui replied:
Okay, done.

djjuhasz commented on internal/workflow/processing.go:
```diff
@@ -246,31 +246,23 @@ func (w *ProcessingWorkflow) Execute(ctx temporalsdk_workflow.Context, req *inge
+			if err := w.deleteOriginalSIP(ctx, state); err != nil {
+				w.logger.Error("Failed to delete original SIP", "err", err.Error())
+			}
+		} else if req.CompletedDir != "" {
```
Is `req.CompletedDir` really mutually exclusive from having a retention period? There's no note in the `ProcessingWorkflowRequest` field comments to indicate that you can't set both `RetentionPeriod` and `CompletedDir`.  Maybe the SIP should be moved to the completed dir (if set), and then the retention period should be applied regardless of whether the SIP is in the completed dir or the original dir?

djjuhasz replied:
Ah, I've traced the `ProcessingWorkflowRequest.CompletedDir` back to https://github.com/artefactual-sdps/enduro/blob/main/internal/watcher/config.go#L83 and it is checked for mutually exclusivity with `RetentionPeriod` at https://github.com/artefactual-sdps/enduro/blob/main/internal/watcher/filesystem.go#L57.

It's really hard to trace the path from `req.CompletedDir` back to `watcher.FilesystemConfig` though, and I think it's a mistake to be passing `req.CompletedDir` when it's an filesystem watcher implementation specific field that we aren't using anywhere else.  Maybe it was necessary to pass `CompletedDir` in the past because we didn't have access to the watcher config directly, but now we have access to it both through the watcher service (https://github.com/artefactual-sdps/enduro/blob/main/internal/watcher/watcher.go#L43) AND the workflow config (https://github.com/artefactual-sdps/enduro/blob/main/internal/workflow/processing.go#L46). :(

djjuhasz commented on enduro.toml:
```diff
@@ -243,6 +248,11 @@ passphrase = ""
+# successful ingest. Do not set to disable automatic deletion of SIPs. Set to "0"
+# to delete uploaded SIPs immediately after they have been ingested.
+# It must be a string format compatible with https://pkg.go.dev/time#ParseDuration.
+retentionPeriod = "0"
```
I don't think it's necessary or even desirable to have a retention period for uploaded SIPs — I don't think there is any way for a system operator to access the SIP in the internal storage location, so I don't think there's a reason to keep it after processing is completed.  Because the SIPs in internal storage can't be seen through the UI, any SIPs stored in internal storage end up being "hidden" SIPs that just take up space.

jraddaoui replied:
Good point. Also, the behavior for the failed SIPs is different already for internal uploads, instead of copying the SIP as failed SIP in the internal bucket we "rename" it to avoid duplication. Following what @djjuhasz said, we could also delete the original in case of failure. However, we never experienced with workflow retries where it may be good to keep a copy to avoid the re-upload.

In any case, I was just following the [issue notes](https://github.com/artefactual-sdps/enduro/issues/1355#issuecomment-3309310033), so I'd love to hear from @fiver-watson before changing this.

fiver-watson replied:
**Team discussion outcome, 2025-10-01**: 

* Aim for consistency - leave retention periods for all upload methods
* Set default behavior when no retention period is configured to be same as "0" - i.e. delete immediately upon success
* In our default config, keep retention period for internal bucket to 0

In the docs, we will add a big warning about setting a retention period for the internal upload space, mentioning that it is also the processing space, and if full, nothing can be ingested.

djjuhasz commented on internal/ingest/goa.go:
```diff
@@ -96,11 +96,12 @@ func (w *goaWrapper) AddSip(ctx context.Context, payload *goaingest.AddSipPayloa
+		SIPName:         s.Name,
+		Type:            enums.WorkflowTypeCreateAip,
+		Key:             payload.Key,
+		RetentionPeriod: w.sipSource.RetentionPeriod(),
```
It seems very awkward to me how we are passing the `RetentionPeriod` to `InitProcessingWorkflow` even though we already have access to the configured value through the `ProcessingWorkflow.cfg`.  I think this is a result of how watchers are handled, where each watcher calls `InitProcessingWorkflow` with it's specific `RetentionPeriod`, but it feels like we are duplicating the information unnecessarily.

I don't think it's feasible in this PR to change how we manage and share configuration data from all the various places a SIP can be ingested, but I think it's something we should consider improving in the future.

jraddaoui replied:
Sounds good, I think we didn't pass the entire config to the workflow initially. And I think in some cases it may be better to just pass the value than looking for it in workflow.

djjuhasz reviewed (commented):
@jraddaoui looks good, but I had a few questions, a few recommendations and I am just finding it confusing how the relationships between the various SIP ingest locations and how we are starting the ingest workflow are managed.

jraddaoui commented:
@djjuhasz, I think I have addressed all you feedback, including the default value change we discussed with the team. Still lacks documentation, and the workflow tests I added are based on https://github.com/artefactual-sdps/enduro/pull/1381, so we'll to merge that one first. But this is ready for another look when you have some time.

djjuhasz reviewed (approved):
LGTM 👍

jraddaoui commented:
Thanks @djjuhasz! But we need to merge https://github.com/artefactual-sdps/enduro/pull/1381 first and update the tests here to use the changes we made there (`defaultParams()`, etc). That one is still waiting for another look ;)

jraddaoui committed:
Add retention period to all SIP origin configs

Allow setting a retention period to the internal uploads and the SIP
sources, and update the current behavior in watched buckets for
consistency. The retention period will default to "0" (automatic
deletion) and has to be set to a negative value to disable the deletion
of the original SIP after a successful ingest. Record a workflow task to
show details in the API/dashboard.

[skip-codecov]

jraddaoui commented:
Thanks again @djjuhasz, I'll merge this one and add the documentation in another PR.
